### PR TITLE
short circuit simple mkString

### PR DESCRIPTION
**Pull request by isaacl**
mkString fast path for 0 or 1 elts

### CODE REVIEW

**Comment by isaacl** [comment]
looks like lazylazy expects `mkString(sep)` calls into `mkString(start, sep, end)`.  whyyyyy

https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/immutable/LazyList.scala#L448-L455

**Comment by isaacl** [comment]
@szeiger can you offer any insight?

**Comment by Ichoran** [comment]
The single and no-args variants of `mkString` are meant to be syntactic sugar for the three-arg variant.  Why not optimize the three-arg one?  That the others aren't `final` seems questionable to me.

**Comment by isaacl** [comment]
the optimization works because the result is equal to the first string. It won’t work if there’s a beginning or end string, which presumably is the reason someone would use the 3 arg variant.

**Comment by Ichoran** [comment]
@isaacl - It will work if the beginning and end strings are empty.

**Comment by SethTisue** [comment]
> That the others aren't final seems questionable to me

I agree they should be made final, and there is still time before M5 to do so.

**Comment by isaacl** [comment]
is this closer to what you're looking for?

**Comment by Ichoran** [comment]
Yes, that looks more like it!  Though I'd manually inline a few more of the empty cases in case the JIT compiler doesn't manage it.  (E.g. rather than doing `start + end` if `start` is nonempty, I'd also check whether `end` was empty.)

**Comment by isaacl** [comment]
alternatively we could call the other way, which avoids checks on start and end for simple mkString calls.

```scala
def mkString(start, sep, end) =
  if (start.length + end.length == 0) mkString(sep)
  else addString(new StringBuilder(), start, sep, end).toString

def mkString(sep) = {
  val it = iterator
  if (it.hasNext) {
    val first = String.valueOf(it.next)
    if (it.hasNext) it.addString(new StringBuilder(first), sep, sep, "").toString
    else first
} else ""
```

**Comment by Ichoran** [comment]
@isaacl - That is cleaner, but it makes overriding particularly difficult to get right.

**Comment by isaacl** [comment]
why should clients override mkString? They can override addString instead

File: src/library/scala/collection/IterableOnce.scala [hunk 1]
```diff
@@ -929,8 +929,19 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
     *
     *  @example  `List(1, 2, 3).mkString("(", "; ", ")") = "(1; 2; 3)"`
     */
-  def mkString(start: String, sep: String, end: String): String =
-    addString(new StringBuilder(), start, sep, end).toString
+  def mkString(start: String, sep: String, end: String): String = {
+    val it = iterator
+    if (!it.hasNext) {
+      if (start.length == 0) end
```
Comment (mkeskells) [inline src/library/scala/collection/IterableOnce.scala:935]: you could optimise for start and end
```
if (start.isEmpty) end
else if (end.isEmpty) start
else start + end
```

**Comment by isaacl** [comment]
Assuming you want to keep all 3 mkString and all 3 addString, seems all mkString variants and the two sugar addString could be marked final.

Overriders should be overriding the 3 arg addString variant, if they only override mkString they're asking for trouble.

**Comment by Ichoran** [comment]
Agreed.  I'm not quite sure what should be final, but leaving all six non-final is kind of ridiculous.  On the other hand, it would be a shame to rule out highly efficient overrides of common special cases.  I can't think of why that would be the case for anything but `mkString` though (where it could be replaced by the identity in case of, for example, `String`).

**Comment by isaacl** [comment]
I'm not convinced we need the two addString sugar methods but no real reason to break things I suppose.

**Comment by isaacl** [comment]
closing in favor of a more comprehensive PR #6878

**Pull request closed without merging**